Persist the config changes of usermanagement

sharidas opened the pull request:
Persist the changes made to usermanagemnt like:
- Show Languages
- Show last login
- Show user backend
- Show storage path

Signed-off-by: Sujith Haridasan <sujith.h@gmail.com>

sharidas commented:
### Summary

In the user management page, the settings to show languages, last login etc were not saved in the db.

### Description

The idea behind this PR is to save the changes made to settings in the user management page for:
- `Show Languages`
- `Show last login`
- `Show user backend`
- `Show store path`

The only option that was being saved was `Send email to new user`. And hence prior to this change set the changes made by the user were lost after relogin. This change set however would help the user regain the configuration which was set.

### How was it tested?

- Create `admin` user
- Enable/Disable any of the options like `Show Languages`, `Show last login`, `Show user backend`, `Show storage path`
- Relogin again to see if these options are same as before.
- It worked as expected.

skjnldsv commented on apps/settings/src/views/Users.vue:
```diff
@@ -235,27 +235,67 @@ export default {
-				this.setLocalStorage('showStoragePath', status)
+			get() { return this.settings.showStoragePath },
+			async set(value) {
+				try {
+					this.$store.commit('setServerData', {
+						...this.settings,
+						showStoragePath: value,
+					})
+					await axios.post(generateUrl('/settings/users/preferences/showstoragepath'), { value: value ? 'yes' : 'no' })
+				} catch (e) {
+					console.error('could not update last login preference: ' + e.message, e)
+				} finally {
+					this.showConfig.showStoragePath = false
+				}
```
This should most likely go into its own function, so you don't duplicate code :)

sharidasan committed:
Persist the config changes of usermanagement

Persist the changes made to usermanagemnt like:
- Show Languages
- Show last login
- Show user backend
- Show storage path

Signed-off-by: Sujith Haridasan <sujith.h@gmail.com>

skjnldsv commented on apps/settings/src/views/Users.vue:
```diff
@@ -360,14 +349,32 @@ export default {
 		},
+		updateSettings(value, key, addToURL) {
+			try {
+				this.loadingSendMail = true
```
Variable name isn't relevant anymore :)

sharidas replied:
Ouch that was a copy/paste error. Thanks for spotting...

skjnldsv commented on apps/settings/src/views/Users.vue:
```diff
@@ -360,14 +349,32 @@ export default {
+			} catch (e) {
+				console.error('could not update ' + key + ' preference: ' + e.message, e)
+			} finally {
+				if (key === 'newUserSendEmail') {
+					this.loadingSendMail = false
```
loadingSendMail will stay to true if not newUserSendEmail then :thinking:

sharidas replied:
Same here ...

sharidas commented:
`make build-js-production` looks clean in my instance. The files mentioned in `Node / 12.x` github action are already updated in the PR. Not sure what is causing the failure for `Node / 12.x`...

skjnldsv commented:
> `make build-js-production` looks clean in my instance. The files mentioned in `Node / 12.x` github action are already updated in the PR. Not sure what is causing the failure for `Node / 12.x`...

do you `npm install` or `npm ci` ?
You need to use `npm ci`

sharidas commented:
> > `make build-js-production` looks clean in my instance. The files mentioned in `Node / 12.x` github action are already updated in the PR. Not sure what is causing the failure for `Node / 12.x`...
> 
> do you `npm install` or `npm ci` ?
> You need to use `npm ci`

Yah I did. Here is the log captured from my console:

```
nextcloud2 git:(save-config-user-management) ✗ npm ci       
npm WARN prepare removing existing node_modules/ before installation

> fsevents@1.2.13 install /home/sujith/test/nextcloud2/node_modules/fsevents
> node install.js


Skipping 'fsevents' build as platform linux is not supported

> core-js@3.6.5 postinstall /home/sujith/test/nextcloud2/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon: 
> https://opencollective.com/core-js 
> https://www.patreon.com/zloirock 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> core-js@2.6.11 postinstall /home/sujith/test/nextcloud2/node_modules/@babel/polyfill/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> core-js@3.6.4 postinstall /home/sujith/test/nextcloud2/node_modules/@nextcloud/moment/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> core-js@2.6.11 postinstall /home/sujith/test/nextcloud2/node_modules/babel-runtime/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> node-sass@4.14.1 install /home/sujith/test/nextcloud2/node_modules/node-sass
> node scripts/install.js

Cached binary found at /home/sujith/.npm/_cacache/node-sass/4.14.1/linux-x64-72_binding.node

> node-sass@4.14.1 postinstall /home/sujith/test/nextcloud2/node_modules/node-sass
> node scripts/build.js

Binary found at /home/sujith/test/nextcloud2/node_modules/node-sass/vendor/linux-x64-72/binding.node
Testing binary
Binary is fine

> core-js@2.6.11 postinstall /home/sujith/test/nextcloud2/node_modules/@babel/runtime-corejs2/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> nodent-runtime@3.2.1 install /home/sujith/test/nextcloud2/node_modules/nodent-runtime
> node build.js

## Built /home/sujith/test/nextcloud2/node_modules/nodent-runtime/dist/index.js
added 1157 packages in 9.404s
➜  nextcloud2 git:(save-config-user-management) ✗
```

skjnldsv commented:
yeah, there is far too much changes, there should not be that many.
Let me rebase and we'll use the compile bot

skjnldsv commented on apps/settings/src/views/Users.vue:
```diff
@@ -360,14 +349,34 @@ export default {
 		},
+		updateSettings(value, key, addToURL) {
+			try {
+				if (key === 'newUserSendEmail') {
+					this.loadingSendMail = true
+				}
```
Well, we should have a loading feedback as well, not only for newUserSendEmail

sharidas replied:
@skjnldsv A loading feedback for `Show Languages`, `Show last login`, `Show user backend` and `Show storage path`? Just to increase my knowledge, I have a query if that's the case. May I know how the loading feedback for the check boxes / display details, would be useful in this case? Am I missing some context here? Don't take me wrongly, pls.

skjnldsv replied:
checkbox should be disabled to prevent sending multiple requests and race conditions.
You can use `loading` and disable all the checkboxes while loading if you want  :)
Might be easier and good enough! :+1:

skjnldsv commented on apps/settings/src/views/Users.vue:
```diff
@@ -360,14 +349,34 @@ export default {
+				})
+				axios.post(generateUrl('/settings/users/preferences/' + addToURL), { value: value ? 'yes' : 'no' })
+			} catch (e) {
+				console.error('could not update ' + key + ' preference: ' + e.message, e)
```
If it fails, we should revert the checkbox to its previous state :)

skjnldsv commented on apps/settings/src/views/Users.vue:
```diff
@@ -360,14 +357,37 @@ export default {
+			} finally {
+				this[loadingData] = false
+
+				if (!shouldUpdate) {
+					// An error occured so lets fall back to previous state
+					document.getElementById(key).checked = value
```
Why not put it in the catch?

sharidas replied:
Done. Moved to catch.

ChristophWurst commented on apps/settings/lib/Controller/UsersController.php:
```diff
@@ -260,6 +260,10 @@ public function usersList() {
 		$serverData['newUserGenerateUserID'] = $this->config->getAppValue('core', 'newUser.generateUserID', 'no') === 'yes';
 		$serverData['newUserRequireEmail'] = $this->config->getAppValue('core', 'newUser.requireEmail', 'no') === 'yes';
 		$serverData['newUserSendEmail'] = $this->config->getAppValue('core', 'newUser.sendEmail', 'yes') === 'yes';
+		$serverData['showLastLogin'] = $this->config->getAppValue('core', 'showlastlogin', 'no') === 'yes';
```
they key `showlastlogin` is way too generic. we should prefix it with something like `user-management`

sharidas replied:
Ok. I will prefix the newly created `showlastlogin`, `showlanguages`, `showuserbackend` and `showstoragepath` with `users`. Prefixing with `user-management` would be too lengthy...

skjnldsv replied:
Or we can use initial states :)

sharidas replied:
Updated with `users`, prefix. Ouch @skjnldsv I saw your comment after updating it.

skjnldsv commented:
What is the status here? :)

sharidas commented:
> What is the status here? :)

I have updated the changes requested :) The conflicts need to be resolved in the branch...

skjnldsv commented:
Please rebase to master :)

juliushaertl commented on apps/settings/lib/Controller/UsersController.php:
```diff
@@ -260,6 +260,10 @@ public function usersList() {
 		$serverData['newUserGenerateUserID'] = $this->config->getAppValue('core', 'newUser.generateUserID', 'no') === 'yes';
 		$serverData['newUserRequireEmail'] = $this->config->getAppValue('core', 'newUser.requireEmail', 'no') === 'yes';
 		$serverData['newUserSendEmail'] = $this->config->getAppValue('core', 'newUser.sendEmail', 'yes') === 'yes';
+		$serverData['showLastLogin'] = $this->config->getAppValue('core', 'users-showlastlogin', 'no') === 'yes';
```
I'd prefer to have this stored in the 'settings' app instead of 'core'

juliushaertl commented on apps/settings/lib/Controller/UsersController.php:
```diff
@@ -271,12 +275,12 @@ public function usersList() {
 		}
 
-		$this->config->setAppValue('core', $key, $value);
+		$this->config->setAppValue('core', 'users-'.$key, $value);
```
Ideally we could introduce them as user values so that each admin user has their own setting for `'showlastlogin', 'showlanguages', 'showuserbackend', 'showstoragepath'` using setUserValue instead

juliushaertl replied:
Also wouldn't this basically break the existing newUser.sendMail setting since the key would now be prefixed with 'users-' ?

juliusknorr reviewed (changes requested):
See inline :wink:

J0WI commented:
:ping_pong: @sharidas
